google/cloud/storage/testbench/gcs_bucket.py

```diff
@@ -499,3 +503,125 @@ def lock_retention_policy(self, request):
+            'object_name': metadata.get('name'),
+            'media': '',
+            'request': request,
+            'done': False,
+        }
+        upload_id = base64.b64encode(metadata.get('name'))
+        self.resumable_uploads[upload_id] = upload
+        location = '%s?uploadType=resumable&upload_id=%s' % (
+            upload_url, upload_id)
+        response = flask.make_response('')
+        response.headers['Location'] = location
+        return response
+
+    def receive_upload_chunk(self, gcs_url, request):
+        """Receive a new upload chunk.
+
+        :param gcs_url: str the base URL for the service.
+        :param request: flask.Request the original http request.
+        :return: the HTTP response.
+        """
+        upload_id = request.args.get('upload_id')
+        if upload_id is None:
+            raise error_response.ErrorResponse(
+                'Missing upload_id in resumable_upload_chunk', status_code=400)
+        upload = self.resumable_uploads.get(upload_id)
+        if upload is None:
+            raise error_response.ErrorResponse(
+                'Cannot find resumable upload %s' % upload_id, status_code=404)
+        # Be gracious in what you accept, if the Content-Range header is not
+        # set we assume it is a good header and it is the end of the file.
+        received_bytes = upload['received_bytes']
+        begin = received_bytes
+        end = received_bytes + len(request.data)
+        total = end
+        content_range = request.headers.get('content-range')
+        if content_range is not None:
+            if content_range.starts_with('*/'):
+                # This is just a query to resume an upload, if it is done, return
+                # an empty
+                response = flask.make_response('')
+                response.headers['Range'] = 'bytes=0-%d' % upload.get('received_bytes')
+                if upload.get('done', False):
+                    # We could also return 200, but this is a better test for the
+                    # library.
+                    response.status_code = 201
+                else:
+                    response.status_code = 308
+                return response
+            match = re.match('([0-9]+)-([0-9]+)/(\*|[0-9]+)', content_range)
+            if not match:
+                raise error_response.ErrorResponse(
+                    'Invalid Content-Range in upload %s' % content_range,
+                    status_code=400)
+            begin = int(match.group(1))
+            end = int(match.group(2))
+            total = int(match.group(3))
+
+        if begin != received_bytes:
+            raise error_response.ErrorResponse(
+                'Mismatched data range, expected data at %d, got %d' % (
+                    received_bytes, begin), status_code=400)
+        if end - begin != len(request.data):
+            raise error_response.ErrorResponse(
+                'Mismatched data range vs. content-length %d, got %d' % (
```

Comment:
Nit: Seeing this error message in a response, I might not know which number goes with which thing.  This might be clearer if written as:

```
'Mismatched data range (%d) vs. content-length (%d)' % (end - begin, len(request.data))
```

Reply:
Done.